pkg/merge/example_deferred_test.go

```diff
@@ -0,0 +1,182 @@
+package merge
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cloudposse/atmos/pkg/schema"
+)
+
+// ExampleMergeWithDeferred demonstrates the complete deferred merge workflow.
+// This example shows how YAML functions are detected, deferred during merge,
+// and then applied after merging to avoid type conflicts.
+func ExampleMergeWithDeferred() {
+	// Create Atmos configuration with replace strategy.
+	cfg := &schema.AtmosConfiguration{
+		Settings: schema.AtmosSettings{
+			ListMergeStrategy: "replace",
+		},
+	}
+
+	// Simulate configuration from multiple imports.
+	// In a real scenario, these would come from different YAML files
+	// in an import chain (e.g., catalog/base.yaml, stacks/dev.yaml, stacks/prod.yaml).
+	inputs := []map[string]any{
+		{
+			// Base catalog with YAML function.
+			"template_config": "!template '{{ .settings.base }}'",
+			"regular_value":   "from-base",
+			"nested": map[string]interface{}{
+				"yaml_func": "!terraform.output vpc.id",
+				"static":    "base-static",
+			},
+		},
+		{
+			// Dev environment override.
+			"template_config": "!template '{{ .settings.dev }}'",
+			"regular_value":   "from-dev",
+			"nested": map[string]interface{}{
+				"static": "dev-static",
+			},
+		},
+		{
+			// Prod environment with concrete value.
+			// Without deferred merge, this would cause a type conflict
+			// if the YAML function returns a different type.
+			"regular_value": "from-prod",
+		},
+	}
+
+	// Perform merge with deferred YAML functions.
+	result, dctx, err := MergeWithDeferred(cfg, inputs)
+	if err != nil {
+		fmt.Printf("Merge error: %v\n", err)
+		return
+	}
+
+	// Show what was deferred (sorted for consistent output).
+	fmt.Println("Deferred YAML functions:")
+	deferredValues := dctx.GetDeferredValues()
+	paths := make([]string, 0, len(deferredValues))
+	for path := range deferredValues {
+		paths = append(paths, path)
+	}
+	sort.Strings(paths)
+	for _, path := range paths {
+		values := deferredValues[path]
+		fmt.Printf("  %s: %d values\n", path, len(values))
+		for i, v := range values {
+			fmt.Printf("    [%d] precedence=%d value=%v\n", i, v.Precedence, v.Value)
+		}
+	}
+
+	// Show merged result (with nil placeholders for YAML functions).
+	fmt.Println("\nMerged result (before applying deferred):")
+	fmt.Printf("  template_config: %v\n", result["template_config"])
+	fmt.Printf("  regular_value: %v\n", result["regular_value"])
+	if nested, ok := result["nested"].(map[string]interface{}); ok {
+		fmt.Printf("  nested.yaml_func: %v\n", nested["yaml_func"])
+		fmt.Printf("  nested.static: %v\n", nested["static"])
+	}
+
+	// Apply deferred merges.
+	// Note: In production, this would process YAML functions.
+	// For this example, they remain as strings since processing is TODO.
+	err = ApplyDeferredMerges(dctx, result, cfg)
+	if err != nil {
+		fmt.Printf("Apply error: %v\n", err)
+		return
+	}
+
+	fmt.Println("\nFinal result (after applying deferred):")
```

Comment:
Use logging or UI layer

Reply:
This is a Go example test (https://go.dev/blog/examples), which serves as executable documentation.

Example tests MUST use `fmt.Println/Printf` because:

1. Their output is captured and compared against the `// Output:` comment for test validation
2. They appear in godoc and pkg.go.dev documentation with their output
3. They demonstrate API usage to developers reading the docs
4. The `fmt` output is part of the test assertion mechanism

Example tests are explicitly exempt from the "use logging/UI layer" rule because they are documentation, not production code. The naming convention `Example*` signals this special purpose.